### PR TITLE
excution and launch plan command added

### PR DESCRIPTION
**Pull request by evalsocket**
Command Added 
- Get Launch Plan
- Get Version 

Basic Unit Test Added for all commands

### CODE REVIEW

**Review by EngHabu — DISMISSED** [review]

File: cmd/get/execution.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		Id : &admin.NamedEntityIdentifier{
+			Project: config.GetConfig().Project,
+			Domain:  config.GetConfig().Domain,
+			Name:    args[0],
+		},
+	})
+	if err != nil {
+		return err
+	}
+	logger.Debugf(ctx, "Retrieved %v excution", len(excution.Executions))
```
Comment (kumare3) [inline cmd/get/execution.go:51]: typo

File: cmd/get/execution.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			return results,err
+		}
+		return results,nil
+	}
+	excution, err := cmdCtx.AdminClient().ListExecutions(ctx, &admin.ResourceListRequest{
+		Limit: 10,
+		Id : &admin.NamedEntityIdentifier{
+			Project: config.GetConfig().Project,
+			Domain:  config.GetConfig().Domain,
+			Name:    args[0],
```
Comment (kumare3) [inline cmd/get/execution.go:45]: what if the name is not provided? should you not be checking that before hand?

File: cmd/get/launch_plan.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			return results,err
+		}
+		return results,nil
+	}
+	launchPlan, err := cmdCtx.AdminClient().ListLaunchPlans(ctx, &admin.ResourceListRequest{
+		Limit: 10,
+		Id : &admin.NamedEntityIdentifier{
+			Project: config.GetConfig().Project,
+			Domain:  config.GetConfig().Domain,
+			Name:    args[0],
```
Comment (kumare3) [inline cmd/get/launch_plan.go:45]: ditto

**Comment by kumare3** [comment]
This looks great. How about unit tests :P for these functions. This will really set a great precedence

**Comment by kumare3** [comment]
Can you rebase with master?

File: cmd/get/execution.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flytestdlib/logger"
+)
+
+// PrintableExcution is the structure for printing Excution
+type PrintableExcution struct {
+	Version          string `header:"Version"`
+	Name             string `header:"Name"`
+	Type             string `header:"Type"`
+	Discoverable     bool   `header:"Discoverable"`
```
Comment (kumare3) [inline cmd/get/execution.go:18]: I dont think execution has discoverable flag?

File: cmd/get/execution_test.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package get
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_transformExcution(t *testing.T) {
```
Comment (kumare3) [inline cmd/get/execution_test.go:10]: typo
```suggestion
func Test_transformExecution(t *testing.T) {
```

File: cmd/get/execution.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+var excutionStructure = map[string]string{
+	"Version":          "$.id.version",
+	"Name":             "$.id.name",
+	"Type":             "$.closure.compiledTask.template.type",
+	"Discoverable":     "$.closure.compiledTask.template.metadata.discoverable",
+	"DiscoveryVersion": "$.closure.compiledTask.template.metadata.discovery_version",
+}
+
+func transformExcution(jsonbody []byte) (interface{}, error) {
```
Comment (kumare3) [inline cmd/get/execution.go:30]:
```suggestion
func transformExecution(jsonbody []byte) (interface{}, error) {
```

File: cmd/get/launch_plan.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flytestdlib/logger"
+)
+
+// PrintableLaunchPlan is the structure for printing Launch Plan
+type PrintableLaunchPlan struct {
+	Version          string `header:"Version"`
+	Name             string `header:"Name"`
+	Type             string `header:"Type"`
+	Discoverable     bool   `header:"Discoverable"`
```
Comment (kumare3) [inline cmd/get/launch_plan.go:18]: i dont think discoverable exists in launchplan

File: cmd/get/launch_plan.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	if err := json.Unmarshal(jsonbody, &results); err != nil {
+		return results, err
+	}
+	return results, nil
+}
+
+func getLaunchPlanFunc(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
+	launchPlanPrinter := printer.Printer{}
+
+	launchPlans, err := cmdCtx.AdminClient().ListLaunchPlans(ctx, &admin.ResourceListRequest{
```
Comment (kumare3) [inline cmd/get/launch_plan.go:41]: can you use the new NamedId lister i created in adminutils

File: pkg/version/version.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package version
```
Comment (kumare3) [inline pkg/version/version.go:1]: we dont need to add this, this is part of `flytestdlib`
https://github.com/lyft/flytestdlib/blob/master/version/version.go
you already have a dependency on flytestdlib

File: cmd/get/launch_plan.go (hunk 8 above)
Comment (kumare3) [reply cmd/get/launch_plan.go:15]: I dont see this change? This is an example of what i mean
https://github.com/lyft/flytectl/blob/master/cmd/get/task.go#L62

Comment (kumare3) [reply cmd/get/launch_plan.go:15]: I think if you follow the set precedence then you can greatly simplify the code

File: cmd/get/project.go [hunk 10]
```diff
@@ -2,42 +2,23 @@ package get
...
-	if err := json.Unmarshal(jsonbody, &results); err != nil {
-		return results, err
-	}
-	return results, nil
-}
-
 func getProjectsFunc(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
 	adminPrinter := printer.Printer{}
-
+	projects, err := cmdCtx.AdminClient().ListProjects(ctx, &admin.ProjectListRequest{})
```
Comment (kumare3) [inline cmd/get/project.go:16]: thank you

File: pkg/version/version.go (hunk 9 above)
Comment (kumare3) [reply pkg/version/version.go:1]: Again I do not see the change? We should remove this file right? and just use the one in flytestdlib

File: pkg/version/version_test.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package version
```
Comment (kumare3) [inline pkg/version/version_test.go:1]: you dont need the test too

File: cmd/get/transform.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
+package get
+
+import (
+	"encoding/json"
+)
+
+func transformExecution(jsonbody []byte) (interface{}, error) {
```
Comment (kumare3) [inline cmd/get/transform.go:7]: why the move to one file?

Comment (evalsocket) [reply cmd/get/transform.go:7]: Is it a bad idea ?

File: cmd/get/execution.go (hunk 1 above)
Comment (evalsocket) [reply cmd/get/execution.go]: fixed

File: cmd/get/execution.go (hunk 2 above)
Comment (evalsocket) [reply cmd/get/execution.go]: fixed

File: cmd/get/launch_plan.go (hunk 3 above)
Comment (evalsocket) [reply cmd/get/launch_plan.go]: fixed

File: cmd/get/execution.go (hunk 4 above)
Comment (evalsocket) [reply cmd/get/execution.go]: fixed

File: cmd/get/execution_test.go (hunk 5 above)
Comment (evalsocket) [reply cmd/get/execution_test.go]: fixed

**Comment by kumare3** [comment]
Looks good, but, it seems you are moving the printer back to `pkg` folder. I purposely moved it to the top folder to match with the new golang conventions

**Comment by evalsocket** [comment]
@kumare3 I am aware of one kind of structure that we follow https://github.com/golang-standards/project-layout. New there is something new then i can move them in root folder

**Review by kumare3 — COMMENTED** [review]
> @kumare3 I am aware of one kind of structure that we follow https://github.com/golang-standards/project-layout. if there is something new then i can move them in root folder

@evalsocket aah I see, we basically are saying that flytectl/pkg/printer.go and others in pkg are shareable? and so we can keep them in pkg. Looks good then

File: cmd/get/workflow.go [hunk 13]
```diff
@@ -2,6 +2,11 @@ package get
 
 import (
 	"context"
+<<<<<<< HEAD
```
Comment (kumare3) [inline cmd/get/workflow.go:5]: merge conflict

File: .goreleaser.yml [hunk 14]
```diff
@@ -11,7 +11,7 @@ builds:
       - windows
       - darwin
     ldflags:
-      - -s -w -X github.com/lyft/flytestdlib/version.Version={{.Version}} -X github.com/lyft/flytestdlib/version.Build={{.ShortCommit}} -X github.com/lyft/flytestdlib/version.BuildTime={{.Date}}
+      - -s -w -X github.com/lyft/flytectl/version.Version={{.Version}} -X github.com/lyft/flytectl/version.Build={{.ShortCommit}} -X github.com/lyft/flytectl/version.BuildTime={{.Date}}
```
Comment (kumare3) [inline .goreleaser.yml:14]: flytestdlib?

**Comment by kumare3** [comment]
merge conflict then looks good to me

File: .goreleaser.yml (hunk 14 above)
Comment (kumare3) [reply .goreleaser.yml:14]: @evalsocket still not changed?

File: cmd/config/config.go [hunk 15]
```diff
@@ -16,16 +16,20 @@ var (
 	section       = config.MustRegisterSection("root", defaultConfig)
 )
 
+// Config hold configration for flytectl flag
 type Config struct {
 	Project string `json:"project" pflag:",Specifies the project to work on."`
 	Domain  string `json:"domain" pflag:",Specified the domain to work on."`
 	Output  string `json:"output" pflag:",Specified the output type."`
+	Filters  string `json:"filters" pflag:",Specified the flyte filter i.e. workflow/execution"`
```
Comment (kumare3) [inline cmd/config/config.go:24]: shouldnt this be a slice?
[]string?
or actually
map[string]string

File: cmd/get/execution.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+		messages = append(messages, m)
+	}
+	return messages
+}
+
+
+
+func getExecutionFunc(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
+	executionPrinter := printer.Printer{}
+	if config.GetConfig().Filters == "" {
```
Comment (kumare3) [inline cmd/get/execution.go:44]: aah this is where you are using Filters. I guess for now, let us just get workflow executions. We can do task executions separately

Lets remove the filters?

File: cmd/config/config.go (hunk 15 above)
Comment (evalsocket) [reply cmd/config/config.go:24]: yes

File: cmd/get/execution.go [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+	{"Name", "$.id.name"},
+	{"LaunchPlan", "$.spec.launchplan.name"},
+	{"Phase", "$.spec.phase"},
+	{"Duration", "$.closure.duration"},
+	{"StartedAt", "$.closure.started_at"},
+	{"Workflow", "$.closure.workflow_id.name"},
+	{"Metadata", "$.spec.metadata"},
+}
+
+var executionSingleColumns = []printer.Column{
```
[the remark this answers is not in the document]
Comment (katrogan) [reply cmd/get/execution.go:23]: is this specifically meant to be the list of columns for single task executions? (a comment would be appreciated in either case :) ) what @kumare3 said holds true, except that the reference launch plan in the ExecutionSpec will actually be an identifier for the reference launch plan

File: cmd/get/execution.go [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+	if err != nil {
+		return err
+	}
+	err = executionPrinter.Print(config.GetConfig().MustOutputFormat(), executionColumns, ExecutionToProtoMessages(excution.Executions)...)
+	if err != nil {
+		return err
+	}
+	return nil
+
+	return nil
```
Comment (katrogan) [inline cmd/get/execution.go:57]: remove duplicate return?

File: cmd/get/execution.go [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+func ExecutionToProtoMessages(l []*admin.Execution) []proto.Message {
+	messages := make([]proto.Message, 0, len(l))
+	for _, m := range l {
+		messages = append(messages, m)
+	}
+	return messages
+}
+
+func getExecutionFunc(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
```
Comment (katrogan) [inline cmd/get/execution.go:39]: nit: looks like you're calling the List endpoint in the method body, can you update the function name (get -> list)?

File: cmd/get/launch_plan.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/lyft/flytectl/pkg/adminutils"
+	"github.com/lyft/flytectl/pkg/printer"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flytestdlib/logger"
+)
+
+var launchplanColumns = []printer.Column{
+	{"Version", "$.id.version"},
+	{"Name", "$.id.name"},
+	{"Type", "$.closure.compiledTask.template.type"},
```
Comment (katrogan) [inline cmd/get/launch_plan.go:17]: can we include launch plan [state](https://github.com/lyft/flyteidl/blob/master/protos/flyteidl/admin/launch_plan.proto#L106) and [schedules](https://github.com/lyft/flyteidl/blob/master/protos/flyteidl/admin/launch_plan.proto#L125) too? this way we don't lose information we have in flyte-cli and it's really useful to get an overview of what's scheduled when listing launch plans

File: cmd/get/launch_plan.go [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+			return err
+		}
+		return nil
+	}
+
+	launchPlans, err := adminutils.GetAllNamedEntities(ctx, cmdCtx.AdminClient().ListLaunchPlanIds, adminutils.ListRequest{Project: config.GetConfig().Project, Domain: config.GetConfig().Domain})
+	if err != nil {
+		return err
+	}
+	logger.Debugf(ctx, "Retrieved %v launch plan", len(launchPlans))
```
Comment (katrogan) [inline cmd/get/launch_plan.go:56]:
```suggestion
	logger.Debugf(ctx, "Retrieved %v launch plans", len(launchPlans))
```

**Comment by evalsocket** [comment]
I removed the execution command and I will create a separate pr for that

**Review by kumare3 — APPROVED** [review]

**Pull request merged**